### PR TITLE
test(calling): add Playwright E2E tests for 2-user call flows

### PR DESCRIPTION
**Pull request by eigengravy**
# COMPLETES https://jira-eng-sjc12.cisco.com/jira/browse/CAI-7739

## This pull request addresses

Playwright E2E tests for the Calling SDK covering 2-user call flows — lifecycle, media, controls, hold, keepalive, errors, and edge cases. Builds on the SDK init + registration suite (#4799) and plugs into the same phased-parallel architecture via a new `SET_2USER` worker.

## Architecture

Extends the existing phased parallel execution model with a new project, `SET_2USER`, that runs 2-party call tests in parallel with the single-user registration sets (`SET_1`, `SET_2`, `SET_3`).

### PROD vs INT account mapping

- **PROD** — `SET_2USER` uses dedicated `USER_4` + `USER_5` accounts, so it runs fully in parallel with `SET_1` / `SET_2` / `SET_3` (which use `USER_1`/`USER_2`/`USER_3`).
- **INT** — `USER_4` / `USER_5` are aliases for the same 3 INT accounts used by the registration sets, so `SET_2USER - INT` declares `dependencies: ['SET_1 - INT', 'SET_2 - INT', 'SET_3 - INT']` to serialize after them and honor the one-account-per-browser constraint.

A `SET_3USER` project scaffold (for transfer flows) is stubbed in the config but commented out — it will be enabled in a follow-up PR (CAI-7741 / transfer tests).

### New test infrastructure
- `playwright/utils/call.ts` — Call action helpers: `makeCall`, `answerCall`, `endCall`, `rejectCall`, `holdCall`, `resumeCall`, `sendDTMF`, `establishCall`, `getMediaStreams`, `cleanupActiveCalls`, plus disconnect/established waiters with Node-side timeout racing (Playwright `waitForFunction` timeouts can be swallowed by WebRTC work on the page)
- `playwright/test-manager.ts` — Extended with a `media: true` setup option (clicks "Get Media Streams" after registration) and now deregisters/closes stale contexts before reuse and on `cleanup()` to prevent leaked backend registrations across serial test groups
- `playwright/test-data.ts` — Added `USER_4`–`USER_6` roles, `getPhoneNumber()` helper, and enabled `SET_2USER` / `SET_3USER` entries in `USER_SETS`
- `playwright.config.ts` — New `SET_2USER - PROD` / `SET_2USER - INT` projects with env-specific dependencies; `SET_3USER` stubs left commented

### Test groups (shared logic, not spec files)
- `call-lifecycle.ts` — Basic 2-party lifecycle + media/disconnect group (fresh contexts per group)
- `call-controls.ts` — Mute/unmute, DTMF, network flap, and hold/resume (including hold+disconnect races and hold/resume API failures via route interception)
- `call-errors.ts` — Connect timeout, ROAP failure, invalid destination, plus edge cases (resume/disconnect race, deregister mid-call, calling unregistered endpoint)
- `call-keepalive.ts` — Call-level keepalive (postStatus) success, 401 teardown, 500 transient recovery, and max-retry exhaustion

### Suite entry point
- `suites/set-2user.spec.ts` — Thin spec file wiring all test groups for the `SET_2USER` worker

### OAuth / fixtures (follow-up fix commit)
- `test-data.ts` — New `REQUIRED_OAUTH_ROLES` list (`USER_1`–`USER_5`) that the OAuth setup validates against
- `utils/oauth.setup.ts` — Fail fast when any required role is missing `_EMAIL` / `_PASSWORD`, instead of only failing on the first account. `USER_6` remains optional until `SET_3USER` ships.

### Techniques used
- Playwright route interception to inject failures (`401` on keepalive, `500` on hold/resume, timeout on call setup, ROAP error responses)
- `page.evaluate()` to inspect `callingClient.getActiveCalls()` state (connected, muted, hold, call counts)
- `context.setOffline(true/false)` to simulate brief network disruption mid-call (CALL-016)
- Per-group fresh browser contexts via `TestManager.beforeAll`; `afterEach` cleans up any lingering active calls and waits a short settle time before the next test
- Node-side `withTimeout` race to guard against Playwright `waitForFunction` timeouts being blocked by WebRTC

### Call Lifecycle (3 tests)
| Test | Description |
|---|---|
| CALL-001 | Outgoing call happy path — full event sequence, both sides connected |
| CALL-002 | Incoming call answer flow — verify callee perspective |
| CALL-003 | Incoming call reject flow — callee rejects, caller receives rejection |

### Call Lifecycle — Media & Disconnect (5 tests)
| Test | Description |
|---|---|
| CALL-014 | Local disconnect — caller hangs up |
| CALL-015 | Remote disconnect — callee hangs up |
| CALL-033 | Unanswered call — caller hangs up after ring timeout |
| CALL-012 | ROAP success — remote audio `srcObject` populated on both sides |
| CALL-018 | Call metrics — `rtp-rxstat` / `rtp-txstat` VQ metrics present during call |

### Call Controls (3 tests)
| Test | Description |
|---|---|
| CALL-030 | Mute and unmute — `isMuted()` toggles correctly |
| CALL-008 | DTMF send — digit sequence delivered during active call |
| CALL-016 | Network flap with active call — call survives brief offline/online toggle |

### Call Hold (3 tests)
| Test | Description |
|---|---|
| CALL-005 | Hold and resume — multiple cycles |
| CALL-019 | Hold then remote disconnect — callee hangs up while caller holds |
| CALL-020 | Hold then immediate disconnect — caller holds and immediately hangs up |

### Call Hold Errors (2 tests)
| Test | Description |
|---|---|
| CALL-006 | Hold API failure — `hold_error` event emitted (route-injected 500) |
| CALL-007 | Resume API failure — `resume_error` event emitted (route-injected 500) |

### Call Keepalive (4 tests)
| Test | Description |
|---|---|
| CALL-024 | Keepalive success — `postStatus` 200 keeps call alive |
| CALL-025 | Keepalive 401 — expired token tears down call |
| CALL-026 | Keepalive 500 with retry — transient failure then recovery |
| CALL-027 | Keepalive max retries exhausted — all status POSTs fail |

### Call Errors (3 tests)
| Test | Description |
|---|---|
| CALL-011 | Call connect timeout — no stuck call on setup failure |
| CALL-013 | ROAP error — media negotiation failure triggers teardown |
| CALL-028 | Call to invalid destination — call fails and cleans up |

### Call Edge Cases (3 tests)
| Test | Description |
|---|---|
| CALL-021 | Resume and disconnect race — concurrent operations on both endpoints |
| CALL-034 | Deregister during active call — callee deregisters mid-call |
| CALL-035 | Call to unregistered endpoint — call fails and cleans up |

**Total: 26 call tests across 8 serial groups, all running in the `SET_2USER` worker.**

image

## The GAI Coding Policy And Copyright Annotation Best Practices ##

- [ ] GAI was not used (or, no additional notation is required)
- [ ] Code was generated entirely by GAI
- [x] GAI was used to create a draft that was subsequently customized or modified
- [ ] Coder created a draft manually that was non-substantively modified by GAI (e.g., refactoring was performed by GAI on manually written code)
- [x] Tool used for AI assistance (GitHub Copilot / Other - specify)
  - [ ] Github Copilot
  - [x] Other - Claude Code
- [x] This PR is related to
  - [ ] Feature
  - [ ] Defect fix
  - [ ] Tech Debt
  - [x] Automation

### I certified that

- [x] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [x] I discussed changes with code owners prior to submitting this pull request
- [x] I have not skipped any automated checks
- [x] All existing and new tests passed
- [x] I have updated the documentation accordingly

### CODE REVIEW

**Review by bhabalan — COMMENTED** [review]
_(Retracted — posted in error. Please disregard; a proper draft with inline comments will follow.)_

**Review by bhabalan — COMMENTED** [review]
disregard this

File: packages/calling/playwright/utils/call.ts [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+ * Race a promise against a Node.js-side timeout.
+ * Playwright's page.waitForFunction timeout sometimes doesn't fire
+ * (possibly due to WebRTC/media operations blocking the browser),
+ * so we enforce an independent timeout from the Node.js side.
+ */
+const withTimeout = <T>(promise: Promise<T>, ms: number, label: string): Promise<T> =>
+  Promise.race([
+    promise,
+    new Promise<never>((_, reject) => {
+      setTimeout(() => reject(new Error(`${label} timed out after ${ms}ms`)), ms);
```
Comment (bhabalan) [inline packages/calling/playwright/utils/call.ts]: The fallback `setTimeout` never gets cleared when the main promise wins the race. Every successful path through `withTimeout` leaves a dangling timer handle (up to 60s for the `establishCall` timeout). Across 26+ tests that adds up and can keep the worker process alive longer than necessary.

Quick fix:

```ts
const withTimeout = <T>(promise: Promise<T>, ms: number, label: string): Promise<T> => {
  let timer: ReturnType<typeof setTimeout>;
  return Promise.race([
    promise.finally(() => clearTimeout(timer)),
    new Promise<never>((_, reject) => {
      timer = setTimeout(() => reject(new Error(`${label} timed out after ${ms}ms`)), ms);
    }),
  ]);
};
```

File: packages/calling/playwright/test-data.ts [hunk 2]
```diff
@@ -81,19 +106,19 @@ export const USER_SETS: Record<string, UserSet> = {
...
+    name: 'SET_2USER',
+    accounts: ['USER_4', 'USER_5'],
+    testSuite: 'set-2user.spec.ts',
+  },
+
+  // 3-user transfer tests (PROD — dedicated accounts, parallel with registration)
+  SET_3USER: {
+    name: 'SET_3USER',
+    accounts: ['USER_4', 'USER_5', 'USER_6'],
+    testSuite: 'set-3user.spec.ts',
```
Comment (bhabalan) [inline packages/calling/playwright/test-data.ts:120]: `SET_3USER` is uncommented here but its Playwright projects are still commented out in the config. This pulls `USER_6` into `uniqueRoles` in the OAuth setup, so the login loop will try to process it even though nothing runs those tests yet. At best it wastes time on a no-op login attempt; at worst it'll fail in environments that don't have `USER_6` credentials. I'd keep this entry commented out until the SET_3USER projects are actually enabled.

Comment (eigengravy) [reply packages/calling/playwright/test-data.ts:120]: I will open a separate PR with the transfer consult flows which is the SET_3USER suite.

**Comment by Shreyas281299** [comment]
### Code review

Found 1 issue:

1. **unregisterLine does not wait for deregistration to complete** (bug due to incomplete cleanup pattern)

The `unregisterLine` function only waits for the "Un registering" UI status (line 30), which indicates deregistration has started but not completed. The final status should be "Unregistered". When `setupContext` tears down existing contexts (test-manager.ts lines 115-120), it calls `unregisterLine` but may proceed to close the browser context before backend deregistration completes. This creates a race condition that could lead to stale device registrations.

The test REG-010 (registration-lifecycle.ts lines 165-170) demonstrates the proper pattern: after calling `unregisterLine`, it explicitly waits for "Unregistered" status. This pattern should be followed in cleanup code to prevent registration leaks, especially given the recent fix in commit 46739e4d81 "improve excessive registrations handling."

https://github.com/webex/webex-js-sdk/blob/66f8e72c61691c0538b41353dde35d52f0d2f028/packages/calling/playwright/utils/registration.ts#L27-L35

🤖 Generated with [Claude Code](https://claude.ai/code)

<sub>- If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Review by Shreyas281299 — APPROVED** [review]

**Review by bhabalan — APPROVED** [review]
